torch/_ops.py

```diff
@@ -925,6 +945,94 @@ def tags(self):
     # TODO: add more methods to expose information about input and output arguments
 
 
+class PythonDispatcherOpOverload(OpOverload):
+    def __repr__(self):
+        return "<PythonDispatcherOpOverload(op='{}.{}', overload='{}')>".format(
+            *self._schema.name.split("::"), self._overloadname
+        )
+
+    def _fallthrough_keys(self) -> list[DispatchKey]:
+        # TODO: we should be calling the fallback for these, but a fallthrough is almost close
+        # enough to the fallback in most cases that we care about.
+        _DEFAULT_FALLTHROUGH_KEYS = [
+            DispatchKey.ADInplaceOrView,
+            DispatchKey.BackendSelect,
+            DispatchKey.PythonTLSSnapshot,
+            DispatchKey.PythonDispatcher,
+        ]
+
+        def _may_use_fallthrough_instead_of_fallback(key: DispatchKey):
+            if torch._C._dispatch_has_kernel_for_dispatch_key(self.name(), key):
+                return torch._C._dispatch_kernel_for_dispatch_key_is_fallthrough(
+                    self.name(), key
+                )
+
+            return (
+                key not in self.py_kernels
+                or self.py_kernels[key] is torch.library.fallthrough_kernel
+            )
+
+        return [
+            key
+            for key in _DEFAULT_FALLTHROUGH_KEYS
+            if _may_use_fallthrough_instead_of_fallback(key)
+        ]
+
+    def __call__(self, /, *args, **kwargs):
+        return self._dispatch_in_python(args, kwargs, self._fallthrough_keys())
+
+    def redispatch(self, /, keyset, *args, **kwargs):
+        return self._dispatch_in_python(args, kwargs, self._fallthrough_keys())
+
```

Comment:
At some point I want to try an exercise of "refactor the python dispatcher to look more like the C++ dispatcher". Concretely:
- each operator has an OperatorEntry
- there is a mapping from DispatchKey->kernel on the OperatorEntry
- OperatorEntry has a DispatchKeyExtractor
- registering a fallback as a py_kernel should modify the DispatchKeyExtractor
- one can use the DispatchKeyExtractor to get the next DispatchKey
- DispatchKeys can have fallbacks in Python
- etc